### PR TITLE
Ingest SCoPe classifications of ZTF sources

### PR DESCRIPTION
**Pull request by dmitryduev**
Ingestion script for SCoPe classifications of ZTF sources, generated by @mcoughlin.

### CODE REVIEW

**Review by mcoughlin — APPROVED** [review]
Looks nice Dima. It feels like at this point some kind of "Ingestor" class would be nice, but probably a bit late now.

File: tools/ingest_ztf_source_classifications.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+import multiprocessing as mp
+import os
+import pandas as pd
+import pathlib
+import pymongo
+import traceback
+from tqdm import tqdm
+
+
+import istarmap  # noqa: F401
```
Comment (stefanv) [inline tools/ingest_ztf_source_classifications.py:12]: Why is this necessary?

**Review by stefanv — APPROVED** [review]
Nice PR!

File: tools/ingest_ztf_source_classifications.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+""" load config and secrets """
+config = load_config(config_file="config.yaml")["kowalski"]
+# init db if necessary
+init_db_sync(config=config)
+
+
+def process_file(_file, _collections):
+
+    # connect to MongoDB:
+    mongo = Mongo(
```
[the remark this answers is not in the document]
Comment (dmitryduev) [reply tools/ingest_ztf_source_classifications.py:30]: Yea, @mcoughlin suggests the same -- will refactor all ingest scripts into a general class + subclassing in a future PR. https://github.com/dmitryduev/kowalski/issues/145

File: tools/ingest_ztf_source_classifications.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+        replica_set=config["database"]["replica_set"],
+        username=config["database"]["username"],
+        password=config["database"]["password"],
+        db=config["database"]["db"],
+        verbose=0,
+    )
+
+    try:
+        with h5py.File(_file, "r") as f:
+            # print(f.keys())
```
[the remark this answers is not in the document]
Comment (dmitryduev) [reply tools/ingest_ztf_source_classifications.py:42]: I just have to google this every time I write an ingestion script for hdf5 files, left this deliberately for the future me :)

File: tools/ingest_ztf_source_classifications.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+            predictions = f["preds"][...]
```
[the remark this answers is not in the document]
Comment (dmitryduev) [reply tools/ingest_ztf_source_classifications.py:43]: Huh, that must be some trick figured out a long time ago that just migrates from script to script...

File: tools/ingest_ztf_source_classifications.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+        cls = str(_file).split("/")[-2]
+        if architecture == "xgb":
+            cls = cls.split(".")[1]
+        elif architecture == "dnn":
+            cls = cls.split("-")[0]
+
+        column_names = ["_id", f"{cls}_{architecture}"]
+
+        df = pd.DataFrame(predictions, columns=column_names)
+        df["_id"] = df["_id"].apply(lambda x: int(x))
```
[the remark this answers is not in the document]
Comment (dmitryduev) [reply tools/ingest_ztf_source_classifications.py:56]: Will do with the refactor https://github.com/dmitryduev/kowalski/issues/145.

File: tools/ingest_ztf_source_classifications.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+        docs = df.to_dict(orient="records")
+
+        requests = []
+
+        for doc in docs:
+
+            _id = doc["_id"]
```
[the remark this answers is not in the document]
Comment (dmitryduev) [reply tools/ingest_ztf_source_classifications.py:64]: It's never None, reading in the hdf5 file above would fail otherwise, I think.

File: tools/ingest_ztf_source_classifications.py (hunk 1 above)
Comment (dmitryduev) [reply tools/ingest_ztf_source_classifications.py:12]: `tqdm` used to not work properly with the some of the standard multiprocessing methods (the progress bar would only update when 100% is done; wasn't too long ago, but they may have fixed it by now) - @kmshin1397 found this trick to make it work.

File: tools/ingest_ztf_source_classifications.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (stefanv) [reply tools/ingest_ztf_source_classifications.py:64]: If it's never none, you can simply do `_id = doc.pop("_id")`.

File: tools/ingest_ztf_source_classifications.py (hunk 1 above)
Comment (stefanv) [reply tools/ingest_ztf_source_classifications.py:12]: What I'm curious about is why the import of istarmap is necessary, when it's not used further down.
